pkg/state/testdata/values.7.hcl

```diff
@@ -0,0 +1,39 @@
+helmfile_vars {
```

Comment:
FWIW, I'm a bit concerned about we now have no way to define local variables like `{{ $foo := ... }}".

Similar to private vs public variables in a standard programming language, I often used the `$name` vars as local variables that are not needed to be reused across files.

I believe your motivation for this HCL feature is to effortlessly share and reference some variables across files. The use of HCL for deferred parsing combined with our DAG code solves this. Awesome!

But private vs public distinction is just missing in the picture maybe?

Can we perhaps introduce something like terraform's "locals" and let users use it as a set of local variables that may or may not be exposed as helm values, and use "helmfile_vars" (or "values" as I proposed separately) to expose them?

```hcl
// locals can be shared across hcl files
// however it is not exposed as helm values
locals {
  a = "a"
  b = "b"
  c = "${local.a}${local.b}"
}

// you expose helm values with values blocks
values {
  // expose a local as-is
  a = local.a
  // use a local to render a helm value
  prefiexed_c = "someprefix${local.c}"
}

// values are {"a":"a","prefiexed_c ":"someprefixab"}

// we can even have a way to append the entire hcl object to helm values :)
locals {
  temp_values = tomap({
    c= local.c
    d= "d"
  })
}

append {
  values = locals.temp_values
}

// values are now {"a":"a","prefiexed_c ":"someprefixab", "c":"ab","d":"d"}
```

Reply:
Indeed, I didn't have this private vs public distinction in mind as I was mainly inspired by terraform for this which doesn't really have it either.

```
Similar to private vs public variables in a standard programming language, I often used the $name vars as local variables that are not needed to be reused across files.
```
From what I understand, your goal is to have a local "name" variable for example, and so that it only lives inside the file so that it doesn't have any conflict elsewhere and avoid the var naming complexity ?
	
If we were to do that, I'd definitely keep the variables accessors for all vars which would go against your previous remark (otherwise if someone would set a value named "local", that'd break it)
Please note that the values, from the go templates, are accessed through the .Values.<var> without the specified accessor which only exist in the hcl values world

So this is how I would translate it: 

* locals block would all be evaluated first before anything as they are indenpendent block and can't have any external dependencies

* In practice : 
file1.hcl
```terraform
values {
  a = "a"
  b = "b"
}
```
file2.hcl
```terraform
// I'd need to check how far I could push it, but I'm not sure having multiple locals blocks would be easy to handle in the same file based on your spec
locals {
  c = "a"
  // This wouldn't work as locals cannot be part of the global gaph in this context
  d = "${hv.a} - ${hv.b}" 
  e = local.c // OK
}

values {
  // As we are aware at runtime in which file the variable is through the hcl.Range object
  // That also means that all locals block are evaluated before any `values` block
  // we can inject the corresponding local block reference in a map[string]map[string]*Values (first map indexes the filename, 2nd the variable)
  val = "${local.c} - ${hv.a}"
}


// Regarding that, if you'd want to expose the locals like this, 
// I don't think of a proper way than to put them in their own vars because appending it raw like this could overwrite existing vars

append {
  values = locals.temp_values
}

// Which'd rather look like this somehow
values {
	myLocals = local
}

```

Reply:
Alright, so I just pushed it this way. I'm updating my original post accordingly regarding the specs